tests/Microsoft.DotNet.XHarness.Apple.Tests/AppRunnerTests.cs

```diff
@@ -142,6 +150,119 @@ public async Task RunOnDeviceTest()
+                       It.IsAny<ILog>(),
+                       Capture.In(appOutputLogs),
+                       Capture.In(appOutputLogs),
+                       It.IsAny<TimeSpan>(),
+                       It.IsAny<Dictionary<string, string>?>(),
+                       It.IsAny<int>(),
+                       Capture.In(cancellationTokens)))
+                .Callback(() =>
+                {
+                    // Signal we have started mlaunch
+                    appStarted.SetResult();
+
+                    // When mlaunch gets signalled to shut down, shut down even our fake mlaunch
+                    cancellationTokens.Last().Register(() => mlaunchCompleted.SetResult(new ProcessExecutionResult
+                    {
+                        TimedOut = true,
+                    }));
+                })
+                .Returns(mlaunchCompleted.Task);
+
+            // Act
+            var appRunner = new AppRunner(
+                _processManager.Object,
+                _snapshotReporterFactory,
+                Mock.Of<ICaptureLogFactory>(),
+                deviceLogCapturerFactory.Object,
+                _mainLog.Object,
+                _logs.Object,
+                _helpers.Object);
+
+            var runTask = appRunner.RunApp(
+                appInformation,
+                new TestTargetOs(TestTarget.Device_iOS, null),
+                s_mockDevice,
+                null,
+                timeout: TimeSpan.FromSeconds(30),
+                signalAppEnd: true,
+                Array.Empty<string>(),
+                Array.Empty<(string, string)>());
+
+            // Everything should hang now since we mimicked mlaunch not being able to tell the app quits
+            // We will wait for XHarness to kick off the mlaunch (the app)
+            Assert.False(runTask.IsCompleted);
+            await Task.WhenAny(appStarted.Task, Task.Delay(1000));
+
+            // XHarness should still be running
+            Assert.False(runTask.IsCompleted);
+
+            // mlaunch should be started
+            Assert.True(appStarted.Task.IsCompleted);
+
+            // We will mimick the app writing the end signal
+            var appLog = appOutputLogs.First();
+            appLog.WriteLine(testEndSignal.ToString());
+
+            // AppTester should now complete fine
+            var result = await runTask;
+
+            // Verify
+            Assert.True(result.Succeeded);
+
+            var expectedArgs = $"-setenv=RUN_END_TAG={testEndSignal} " +
+                "--disable-memory-limits " +
+                $"--devname {s_mockDevice.DeviceIdentifier} " +
+                $"--launchdevbundleid {AppBundleIdentifier} " +
+                "--wait-for-exit";
+
+            Assert.Equal(mlaunchArguments.Last().AsCommandLine(), expectedArgs);
+
+            _snapshotReporter.Verify(x => x.StartCaptureAsync(), Times.AtLeastOnce);
+            _snapshotReporter.Verify(x => x.StartCaptureAsync(), Times.AtLeastOnce);
```

Comment:
why two times please?

Reply:
A mistake!

Reply:
Oh, it was supposed to be `StopCapture()` I guess